format/FormatMultiImage.py

```diff
@@ -228,22 +228,23 @@ def get_imageset(
                 indices=single_file_indices,
             )
 
+            if single_file_indices is None:
+                single_file_indices = list(range(format_instance.get_num_images()))
+
             # If any are None then read from format
             if [beam, detector, goniometer, scan].count(None) != 0:
 
                 # Get list of models
-                beam = []
-                detector = []
-                goniometer = []
-                scan = []
-                for i in range(format_instance.get_num_images()):
-                    beam.append(format_instance.get_beam(i))
-                    detector.append(format_instance.get_detector(i))
-                    goniometer.append(format_instance.get_goniometer(i))
-                    scan.append(format_instance.get_scan(i))
-
-            if single_file_indices is None:
-                single_file_indices = list(range(format_instance.get_num_images()))
+                num_images = format_instance.get_num_images()
+                beam = [None for _ in range(num_images)]
+                detector = [None for _ in range(num_images)]
+                goniometer = [None for _ in range(num_images)]
+                scan = [None for _ in range(num_images)]
+                for i in single_file_indices:
+                    beam[i] = format_instance.get_beam(i)
+                    detector[i] = format_instance.get_detector(i)
+                    goniometer[i] = format_instance.get_goniometer(i)
+                    scan[i] = format_instance.get_scan(i)
 
             # Set the list of models
             for i in range(len(single_file_indices)):
```

Comment:
```suggestion
            for i, index in enumerate(single_file_indices):
```
the four lines below would need refactoring, too (which unfortunately I can't put in this suggestion).